Archives/VolFile.cpp

```diff
@@ -389,103 +382,102 @@ namespace Archives
 	}
 
 
-
-
-
 	// Reads a tag in the .vol file and returns the length of that section.
 	// If tagText does not match what is in the file or if the length is 
-	// invalid then this function returns -1.
-	// Note: tagText is a 4 byte field
-	int VolFile::ReadTag(int offset, const char *tagText)
+	// invalid then an error is thrown.
+	uint32_t VolFile::ReadTag(const std::array<char, 4>& tagName)
 	{
-		char *addr = (char*)m_BaseOfFile + offset;
+		// Tags are 4 chars in length and do not include a null terminator
+		std::array<char, 4> tagFromFile;
+		archiveFileReader->Read(tagFromFile.data(), tagFromFile.size());
 
-		if (*(int*)addr != *(int*)tagText) { 	   // Make sure the tag matches
-			return -1;
+		if (tagFromFile != tagName) {
+			throw std::runtime_error("Tag " + std::string(tagFromFile.data(), tagFromFile.size()) + " does not match requested tag of " + std::string(tagName.data(), tagName.size()));
 		}
 
-		int length = *(int*)(addr + 4);			   // Get the length (and length tag)
-		if ((length & 0x80000000) != 0x80000000) { // Check for the tag (MSB set)
-			return -1;
+		uint32_t length = 0;
+		archiveFileReader->Read(&length, sizeof(length));
+
+		// Check for the tag (MSB set)
+		if ((length & 0x80000000) != 0x80000000) {
+			throw std::runtime_error("Tag " + std::string(tagName.data()) + " contains an invalid length.");
```

Comment:
Okay, so I'm looking at the bitwise operation here. 

0x8000 0000 is equivalent to b1000 0000 0000 0000 0000 0000 0000 0000

& will mask out all values not set, so we are checking if the highest order bit is set in a 32 bit (4 byte) field is set to 0 or 1. In this case, if the highest order bit is set to 1, it indicates 2 byte padding is used. While 2 byte padding is mentioned in the volume documentation, it is not actually available for use?

Thanks,
Brett‬

Reply:
The existing VOL files never used 2-byte padding, and my writing code always output 4-byte padding, so I never bothered to support 2-byte padding. I only know about it because the game has code paths to support 2-byte padding.

The length is stored in the lower 31 bits, so technically it's not the length that's invalid. It's just the padding bit setting that is unsupported.